File: dotnet/System/workspace/csharp/adapters/allors.workspace.adapters/session/originstate/RecordBasedOriginState.cs
// <copyright file="RecordBasedOriginState.cs" company="Allors bvba">
// Copyright (c) Allors bvba. All rights reserved.
// Licensed under the LGPL license. See LICENSE file in the project root for full license information.
// </copyright>

namespace Allors.Workspace.Adapters
{
    using System;
    using System.Collections.Generic;
    using System.Diagnostics;
    using System.Linq;
    using Meta;
    using Ranges;

    public abstract class RecordBasedOriginState
    {
        public abstract Strategy Strategy { get; }

        protected bool HasChanges => this.Record == null || this.ChangedRoleByRelationType?.Count > 0;

        protected abstract IEnumerable<IRoleType> RoleTypes { get; }

        protected abstract IRecord Record { get; }

        protected IRecord PreviousRecord { get; set; }

        public Dictionary<IRelationType, object> ChangedRoleByRelationType { get; protected set; }

        private Dictionary<IRelationType, object> PreviousChangedRoleByRelationType { get; set; }

        public object GetUnitRole(IRoleType roleType)
        {
            if (this.ChangedRoleByRelationType != null && this.ChangedRoleByRelationType.TryGetValue(roleType.RelationType, out var role))
            {
                return role;
            }

            return this.Record?.GetRole(roleType);
        }

        public void SetUnitRole(IRoleType roleType, object role) => this.SetChangedRole(roleType, role);

        public Strategy GetCompositeRole(IRoleType roleType)
        {
            if (this.ChangedRoleByRelationType != null && this.ChangedRoleByRelationType.TryGetValue(roleType.RelationType, out var changedRole))
            {
                return (Strategy)changedRole;
            }

            var role = this.Record?.GetRole(roleType);
            return role != null ? this.Session.GetStrategy((long)role) : null;
        }

        public void SetCompositeRole(IRoleType roleType, Strategy role)
        {
            var previousRole = this.GetCompositeRole(roleType);

            if (previousRole == role)
            {
                return;
            }

            var associationType = roleType.AssociationType;
            if (associationType.IsOne && role != null)
            {
                var previousAssociation = this.Session.GetCompositeAssociation(role, associationType);
                this.SetChangedRole(roleType, role);
                if (associationType.IsOne && previousAssociation != null)
                {
                    // OneToOne
                    previousAssociation.SetRole(roleType, null);
                }
            }
            else
            {
                this.SetChangedRole(roleType, role);
            }
        }

        public IEnumerable<Strategy> GetCompositesRole(IRoleType roleType)
        {
            if (this.ChangedRoleByRelationType != null && this.ChangedRoleByRelationType.TryGetValue(roleType.RelationType, out var changedRole))
            {
                return ((ISet<Strategy>)changedRole).ToArray();
            }

            var role = this.Record?.GetRole(roleType);

            if (role is ISet<Strategy>)
            {
                Debugger.Break();
            }

            return role != null ? this.Ranges.Ensure(role).Select(v => this.Session.GetStrategy(v)) : Array.Empty<Strategy>();
        }


        public void AddCompositesRole(IRoleType roleType, Strategy roleToAdd)
        {
            var associationType = roleType.AssociationType;
            var previousAssociation = this.Session.GetCompositeAssociation(roleToAdd, associationType);

            var previousRole = this.GetCompositesRole(roleType);
            var role = previousRole is ISet<Strategy> set ? set : new HashSet<Strategy>(previousRole);

            if (role.Contains(roleToAdd))
            {
                return;
            }

            role.Add(roleToAdd);
            this.SetChangedRole(roleType, role);

            if (associationType.IsMany)
            {
                return;
            }

            // OneToMany
            previousAssociation?.SetRole(roleType, null);
        }

        public void RemoveCompositesRole(IRoleType roleType, Strategy roleToRemove)
        {
            var previousRole = this.GetCompositesRole(roleType);
            var role = previousRole is ISet<Strategy> set ? set : new HashSet<Strategy>(previousRole);

            if (!role.Contains(roleToRemove))
            {
                return;
            }

            role.Remove(roleToRemove);
            this.SetChangedRole(roleType, role);
        }

        public void SetCompositesRole(IRoleType roleType, IEnumerable<Strategy> rolesToSet)
        {
            var previousRole = this.GetCompositesRole(roleType);

            var role = rolesToSet != null ? new HashSet<Strategy>(rolesToSet) : new HashSet<Strategy>();

            this.SetChangedRole(roleType, role);

            var associationType = roleType.AssociationType;
            if (associationType.IsMany)
            {
                return;
            }

            // OneToMany
            foreach (var addedRole in role.Except(previousRole))
            {
                var previousAssociation = this.Session.GetCompositeAssociation(addedRole, associationType);
                previousAssociation?.SetRole(roleType, null);
            }
        }

        public void Checkpoint(ChangeSet changeSet)
        {
            // Same record
            if (this.PreviousRecord == null || this.Record == null || this.Record.Version == this.PreviousRecord.Version)
            {
                // No previous changed roles
                if (this.PreviousChangedRoleByRelationType == null)
                {
                    if (this.ChangedRoleByRelationType != null)
                    {
                        // Changed roles
                        foreach (var kvp in this.ChangedRoleByRelationType)
                        {
                            var relationType = kvp.Key;
                            var current = kvp.Value;
                            var previous = this.Record?.GetRole(relationType.RoleType);
                            if (relationType.RoleType.ObjectType.IsUnit)
                            {
                                changeSet.DiffUnit(this.Strategy, relationType, current, previous);
                            }
                            else
                            {
                                if (relationType.RoleType.IsOne)
                                {
                                    changeSet.DiffComposite(this.Strategy, relationType, (Strategy)current, (long?)previous);
                                }
                                else
                                {
                                    changeSet.DiffComposites(this.Strategy, relationType, (ISet<Strategy>)current, (IRange)previous);
                                }
                            }
                        }
                    }
                }
                // Previous changed roles
                else
                {
                    foreach (var kvp in this.ChangedRoleByRelationType)
                    {
                        var relationType = kvp.Key;
                        var current = kvp.Value;

                        this.PreviousChangedRoleByRelationType.TryGetValue(relationType, out var previous);

                        if (relationType.RoleType.ObjectType.IsUnit)
                        {
                            changeSet.DiffUnit(this.Strategy, relationType, current, previous);
                        }
                        else
                        {
                            if (relationType.RoleType.IsOne)
                            {
                                changeSet.DiffComposite(this.Strategy, relationType, (Strategy)current, (long?)previous);
                            }
                            else
                            {
                                changeSet.DiffComposites(this.Strategy, relationType, (ISet<Strategy>)current, (IRange)previous);
                            }
                        }
                    }
                }
            }
            // Different record
            else
            {
                foreach (var roleType in this.RoleTypes)
                {
                    var relationType = roleType.RelationType;

                    object previous = null;
                    object current = null;

                    if (this.PreviousChangedRoleByRelationType?.TryGetValue(relationType, out previous) == true)
                    {
                        if (this.ChangedRoleByRelationType?.TryGetValue(relationType, out current) != true)
                        {
                            current = this.Record.GetRole(roleType);
                        }

                        if (relationType.RoleType.ObjectType.IsUnit)
                        {
                            changeSet.DiffUnit(this.Strategy, relationType, current, previous);
                        }
                        else
                        {
                            if (relationType.RoleType.IsOne)
                            {
                                changeSet.DiffComposite(this.Strategy, relationType, (Strategy)current, (Strategy)previous);
                            }
                            else
                            {
                                changeSet.DiffComposites(this.Strategy, relationType, (ISet<Strategy>)current, (ISet<Strategy>)previous);
                            }
                        }

                    }
                    else
                    {
                        previous = this.PreviousRecord?.GetRole(roleType);
                        if (this.ChangedRoleByRelationType?.TryGetValue(relationType, out current) != true)
                        {
                            current = this.Record.GetRole(roleType);
                        }

                        if (relationType.RoleType.ObjectType.IsUnit)
                        {
                            changeSet.DiffUnit(this.Strategy, relationType, current, previous);
                        }
                        else
                        {
                            if (relationType.RoleType.IsOne)
                            {
                                changeSet.DiffComposite(this.Strategy, relationType, (long?)current, (Strategy)previous);
                            }
                            else
                            {
                                changeSet.DiffComposites(this.Strategy, relationType, (IRange)current, (ISet<Strategy>)previous);
                            }
                        }
                    }
                }
            }

            this.PreviousRecord = this.Record;
            this.PreviousChangedRoleByRelationType = this.ChangedRoleByRelationType?.ToDictionary(v => v.Key,
                v => v.Value switch
                    {
                        ISet<Strategy> set => new HashSet<Strategy>(set),
                        _ => v.Value
                    });
        }

        public void Diff(IList<IDiff> diffs)
        {
            if (this.ChangedRoleByRelationType == null)
            {
                return;
            }

            var ranges = this.Session.Workspace.Ranges;

            foreach (var kvp in this.ChangedRoleByRelationType)
            {
                var relationType = kvp.Key;
                var roleType = relationType.RoleType;

                var changed = kvp.Value;
                var original = this.Record?.GetRole(roleType);

                if (roleType.ObjectType.IsUnit)
                {
                    diffs.Add(new UnitDiff(relationType, this.Strategy)
                    {
                        OriginalRole = original,
                        ChangedRole = changed,
                    });
                }
                else if (roleType.IsOne)
                {
                    diffs.Add(new CompositeDiff(relationType, this.Strategy)
                    {
                        OriginalRole = original != null ? this.Session.GetStrategy((long)original) : null,
                        ChangedRole = (Strategy)changed,
                    });
                }
                else
                {
                    diffs.Add(new CompositesDiff(relationType, this.Strategy)
                    {
                        OriginalRoles = original != null ? ranges.Ensure(original).Select(v => this.Session.GetStrategy(v)).ToArray() : Array.Empty<Strategy>(),
                        ChangedRoles = ((ISet<Strategy>)changed).ToArray(),
                    });
                }
            }
        }

        public bool CanMerge(IRecord newRecord)
        {
            if (this.ChangedRoleByRelationType == null)
            {
                return true;
            }

            var ranges = this.Session.Workspace.Ranges;

            foreach (var kvp in this.ChangedRoleByRelationType)
            {
                var relationType = kvp.Key;
                var roleType = relationType.RoleType;

                var original = this.Record?.GetRole(roleType);
                var newOriginal = newRecord?.GetRole(roleType);

                if (roleType.ObjectType.IsUnit)
                {
                    if (!Equals(original, newOriginal))
                    {
                        return false;
                    }
                }
                else if (roleType.IsOne)
                {
                    if (!Equals(original, newOriginal))
                    {
                        return false;
                    }
                }
                else if (!ranges.Ensure(original).Equals(ranges.Ensure(newOriginal)))
                {
                    return false;
                }
            }

            return true;
        }

        public void Reset() => this.ChangedRoleByRelationType = null;

        public bool IsAssociationForRole(IRoleType roleType, Strategy forRole)
        {
            if (roleType.IsOne)
            {
                var compositeRole = this.GetCompositeRole(roleType);
                return compositeRole == forRole;
            }

            var compositesRole = this.GetCompositesRole(roleType);
            return compositesRole.Contains(forRole);
        }

        protected abstract void OnChange();

        private void SetChangedRole(IRoleType roleType, object role)
        {
            this.ChangedRoleByRelationType ??= new Dictionary<IRelationType, object>();
            this.ChangedRoleByRelationType[roleType.RelationType] = role;
            this.OnChange();
        }

        #region Proxy Properties

        protected long Id => this.Strategy.Id;

        protected IClass Class => this.Strategy.Class;

        protected Session Session => this.Strategy.Session;

        protected Workspace Workspace => this.Session.Workspace;

        private IRanges Ranges => this.Strategy.Session.Workspace.Ranges;

        #endregion
    }
}

File: typescript/libs/workspace/adapters/system/src/lib/session/originstate/RecordBasedOriginState.ts
import { IRecord } from '../../IRecord';
import { ChangeSet } from '../ChangeSet';
import { Strategy } from '../Strategy';
import { Workspace } from '../../workspace/Workspace';
import { Session } from '../Session';
import { Class, RelationType, RoleType } from '@allors/workspace/meta/system';
import { ICompositeDiff, ICompositesDiff, IDiff, IUnit, IUnitDiff } from '@allors/workspace/domain/system';
import { frozenEmptyArray } from '../../collections/frozenEmptyArray';
import { IRange, Ranges } from '../../collections/ranges/Ranges';

export abstract class RecordBasedOriginState {
  abstract strategy: Strategy;

  protected hasChanges(): boolean {
    return this.record == null || this.changedRoleByRelationType?.size > 0;
  }

  protected abstract roleTypes: Set<RoleType>;

  protected abstract record: IRecord;

  protected previousRecord: IRecord;

  changedRoleByRelationType: Map<RelationType, unknown>;

  private previousChangedRoleByRelationType: Map<RelationType, unknown>;

  getUnitRole(roleType: RoleType): IUnit {
    if (this.changedRoleByRelationType != null && this.changedRoleByRelationType.has(roleType.relationType)) {
      return this.changedRoleByRelationType.get(roleType.relationType) as IUnit;
    }

    return this.record?.getRole(roleType) as IUnit;
  }

  setUnitRole(roleType: RoleType, role: IUnit) {
    this.setChangedRole(roleType, role);
  }

  getCompositeRole(roleType: RoleType): Strategy {
    if (this.changedRoleByRelationType != null && this.changedRoleByRelationType.has(roleType.relationType)) {
      return this.changedRoleByRelationType.get(roleType.relationType) as Strategy;
    }

    const role = this.record?.getRole(roleType) as number;
    return role != null ? this.session.getStrategy(role) : null;
  }

  setCompositeRole(roleType: RoleType, role?: Strategy) {
    const previousRole = this.getCompositeRole(roleType);

    if (previousRole == role) {
      return;
    }

    const associationType = roleType.associationType;
    if (associationType.isOne && role != null) {
      const previousAssociation = this.session.getCompositeAssociation(role, associationType);
      this.setChangedRole(roleType, role);

      if (associationType.isOne && previousAssociation != null) {
        //  OneToOne
        previousAssociation.setRole(roleType, null);
      }
    } else {
      this.setChangedRole(roleType, role);
    }
  }

  getCompositesRole(roleType: RoleType): IRange<Strategy> {
    if (this.changedRoleByRelationType != null && this.changedRoleByRelationType.has(roleType.relationType)) {
      return this.changedRoleByRelationType.get(roleType.relationType) as IRange<Strategy>;
    }

    const role = this.record?.getRole(roleType) as IRange<number>;
    return role != null ? this.strategyRanges.importFrom(role.map((v) => this.session.getStrategy(v))) : (frozenEmptyArray as Strategy[]);
  }

  addCompositesRole(roleType: RoleType, roleToAdd: Strategy) {
    const associationType = roleType.associationType;
    const previousAssociation = this.session.getCompositeAssociation(roleToAdd, associationType);

    let role = this.getCompositesRole(roleType);

    if (this.strategyRanges.has(role, roleToAdd)) {
      return;
    }

    role = this.strategyRanges.add(role, roleToAdd);
    this.setChangedRole(roleType, role);

    if (associationType.isMany) {
      return;
    }

    //  OneToMany
    previousAssociation?.setRole(roleType, null);
  }

  removeCompositesRole(roleType: RoleType, roleToRemove: Strategy) {
    let role = this.getCompositesRole(roleType);

    if (!this.strategyRanges.has(role, roleToRemove)) {
      return;
    }

    role = this.strategyRanges.remove(role, roleToRemove);
    this.setChangedRole(roleType, role);
  }

  setCompositesRole(roleType: RoleType, role: IRange<Strategy>) {
    const previousRole = this.getCompositesRole(roleType);

    this.setChangedRole(roleType, role);

    const associationType = roleType.associationType;
    if (associationType.isMany) {
      return;
    }

    //  OneToMany
    for (const addedRole of this.strategyRanges.enumerate(this.strategyRanges.difference(role, previousRole))) {
      const previousAssociation = this.session.getCompositeAssociation(addedRole, associationType);
      previousAssociation?.setRole(roleType, null);
    }
  }

  checkpoint(changeSet: ChangeSet) {
    //  Same record
    if (this.previousRecord == null || this.record == null || this.record.version == this.previousRecord.version) {
      //  No previous changed roles
      if (this.previousChangedRoleByRelationType == null) {
        if (this.changedRoleByRelationType != null) {
          //  Changed roles
          this.changedRoleByRelationType.forEach((current, relationType) => {
            const previous = this.record?.getRole(relationType.roleType);

            if (relationType.roleType.objectType.isUnit) {
              changeSet.diffUnit(this.strategy, relationType, current, previous);
            } else if (relationType.roleType.isOne) {
              changeSet.diffCompositeStrategyRecord(this.strategy, relationType, current as Strategy, previous as number);
            } else {
              changeSet.diffCompositesStrategyRecord(this.strategy, relationType, current as IRange<Strategy>, previous as IRange<number>);
            }
          });
        }
      } else {
        this.changedRoleByRelationType.forEach((current, relationType) => {
          const previous = this.previousChangedRoleByRelationType?.get(relationType);

          if (relationType.roleType.objectType.isUnit) {
            changeSet.diffUnit(this.strategy, relationType, current, previous);
          } else if (relationType.roleType.isOne) {
            changeSet.diffCompositeStrategyRecord(this.strategy, relationType, current as Strategy, previous as number);
          } else {
            changeSet.diffCompositesStrategyRecord(this.strategy, relationType, current as IRange<Strategy>, previous as IRange<number>);
          }
        });
      }

      //  Previous changed roles
    } else {
      //  Different record
      this.roleTypes.forEach((roleType) => {
        const relationType = roleType.relationType;

        if (this.previousChangedRoleByRelationType?.has(relationType)) {
          const previous = this.previousChangedRoleByRelationType.get(relationType);
          const current = this.changedRoleByRelationType?.has(relationType) ? this.changedRoleByRelationType.get(relationType) : this.record.getRole(roleType);

          if (relationType.roleType.objectType.isUnit) {
            changeSet.diffUnit(this.strategy, relationType, current, previous);
          } else if (relationType.roleType.isOne) {
            changeSet.diffCompositeStrategyStrategy(this.strategy, relationType, current as Strategy, previous as Strategy);
          } else {
            changeSet.diffCompositesStrategyStrategy(this.strategy, relationType, current as IRange<Strategy>, previous as IRange<Strategy>);
          }
        } else {
          const previous = this.previousRecord?.getRole(roleType);
          const current = this.changedRoleByRelationType?.has(relationType) ? this.changedRoleByRelationType?.get(relationType) : this.record.getRole(roleType);

          if (relationType.roleType.objectType.isUnit) {
            changeSet.diffUnit(this.strategy, relationType, current, previous);
          } else if (relationType.roleType.isOne) {
            changeSet.diffCompositeRecordRecord(this.strategy, relationType, current as number, previous as number);
          } else {
            changeSet.diffCompositesRecordRecord(this.strategy, relationType, current as IRange<number>, previous as IRange<number>);
          }
        }
      });
    }

    this.previousRecord = this.record;
    this.previousChangedRoleByRelationType = this.changedRoleByRelationType;
  }

  diff(diffs: IDiff[]) {
    if (this.changedRoleByRelationType == null) {
      return;
    }

    for (const [relationType, changed] of this.changedRoleByRelationType) {
      const roleType = relationType.roleType;
      const original = this.record?.getRole(roleType);

      if (roleType.objectType.isUnit) {
        const diff: IUnitDiff = {
          relationType,
          assocation: this.strategy,
          originalRole: original as IUnit,
          changedRole: changed as IUnit,
        };

        diffs.push(diff);
      } else if (roleType.isOne) {
        const diff: ICompositeDiff = {
          relationType,
          assocation: this.strategy,
          originalRole: original != null ? this.strategy.session.getStrategy(original as number) : null,
          changedRole: changed as Strategy,
        };

        diffs.push(diff);
      } else {
        const diff: ICompositesDiff = {
          relationType,
          assocation: this.strategy,
          originalRoles: original != null ? (original as IRange<number>)?.map((v) => this.strategy.session.getStrategy(v)) ?? frozenEmptyArray : frozenEmptyArray,
          changedRoles: [...(changed as Set<Strategy>)],
        };

        diffs.push(diff);
      }
    }
  }

  canMerge(newRecord: IRecord): boolean {
    if (this.changedRoleByRelationType == null) {
      return true;
    }

    for (const [relationType] of this.changedRoleByRelationType) {
      const roleType = relationType.roleType;
      const original = this.record?.getRole(roleType);
      const newOriginal = newRecord?.getRole(roleType);

      if (roleType.objectType.isUnit) {
        if (original !== newOriginal) {
          return false;
        }
      } else if (roleType.isOne) {
        if (original !== newOriginal) {
          return false;
        }
      } else if (!this.recordRanges.equals(original as number[], newOriginal as number[])) {
        return false;
      }
    }

    return true;
  }

  reset() {
    this.changedRoleByRelationType = null;
  }

  isAssociationForRole(roleType: RoleType, forRole: Strategy): boolean {
    if (roleType.isOne) {
      const compositeRole = this.getCompositeRole(roleType);
      return compositeRole == forRole;
    }
    const composites = this.getCompositesRole(roleType);
    if (composites != null) {
      for (const role of composites) {
        if (role === forRole) {
          return true;
        }
      }
    }

    return false;
  }

  protected abstract onChange();

  private setChangedRole(roleType: RoleType, role: unknown) {
    this.changedRoleByRelationType ??= new Map<RelationType, unknown>();
    this.changedRoleByRelationType.set(roleType.relationType, role);
    this.onChange();
  }

  //#region Proxy
  protected get id(): number {
    return this.strategy.id;
  }

  protected get class(): Class {
    return this.strategy.cls;
  }

  protected get session(): Session {
    return this.strategy.session;
  }

  protected get workspace(): Workspace {
    return this.strategy.session.workspace;
  }

  protected get strategyRanges(): Ranges<Strategy> {
    return this.strategy.session.ranges;
  }

  protected get recordRanges(): Ranges<number> {
    return this.strategy.session.workspace.ranges;
  }
  //#endregion
}
